File: internal/mail/sniffer_test.go
package mail

import (
	"testing"
)

func TestExtractOTP(t *testing.T) {
	tests := []struct {
		name         string
		subject      string
		body         string
		expectedCode string
		expectedLink string
	}{
		{
			name:         "中文验证码冒号",
			subject:      "欢迎注册",
			body:         "您的验证码是：849201，请在 10 分钟内输入。",
			expectedCode: "849201",
		},
		{
			name:         "英文 verification code",
			subject:      "Your verification code",
			body:         "Here is your verification code: 394821 for login.",
			expectedCode: "394821",
		},
		{
			name:         "英文 security code 短横或空格",
			subject:      "Security Alert",
			body:         "Use security code 9283 to verify your account.",
			expectedCode: "9283",
		},
		{
			name:         "英文 your code is",
			subject:      "Sign in to your account",
			body:         "Your code is: 482019",
			expectedCode: "482019",
		},
		{
			name:         "包含激活链接",
			subject:      "Confirm your email",
			body:         "Click here: https://example.com/confirm?token=xyz123 to activate.",
			expectedLink: "https://example.com/confirm?token=xyz123",
		},
		{
			name:         "跨行且包含干扰词 Your code is",
			subject:      "Your verification code",
			body:         "Your code is: 582910. Do not share it.",
			expectedCode: "582910",
		},
		{
			name:         "5位验证码且包含年份干扰",
			subject:      "2026年注册验证",
			body:         "您的验证口令是 78912",
			expectedCode: "78912",
		},
		{
			name:         "连字号验证码 (123-456)",
			subject:      "Telegram login code",
			body:         "Your login code is 123-456. Do not give this code to anyone.",
			expectedCode: "123456",
		},
		{
			name:         "带空格连字号验证码与 One-time password",
			subject:      "Your One-Time Password",
			body:         "Your one-time password is: 987 654 for verification.",
			expectedCode: "987654",
		},
		{
			name:         "反向倒装且包含修饰定语_中文",
			subject:      "登录确认",
			body:         "849201 为您的本次登录验证码，切勿泄露。",
			expectedCode: "849201",
		},
		{
			name:         "反向倒装且包含服务商品牌修饰_英文",
			subject:      "Security Code",
			body:         "492019 is your AWS verification code. Valid for 10 minutes.",
			expectedCode: "492019",
		},
		{
			name:         "反向倒装5位安全码",
			subject:      "Account PIN",
			body:         "58291 is your security code.",
			expectedCode: "58291",
		},
		{
			name:         "PIN独立关键词提取",
			subject:      "Security",
			body:         "Your PIN is: 739201",
			expectedCode: "739201",
		},
		{
			name:         "OTP独立关键词提取",
			subject:      "Login Authentication",
			body:         "Your OTP is 829104 for verification.",
			expectedCode: "829104",
		},
		{
			name:         "韩文ChatGPT认证邮件",
			subject:      "ChatGPT 인증 코드",
			body:         "다음 임시 인증 코드를 입력해 계속하세요: 576932 ChatGPT 계정을 생성하고자 하는 것이 본인이 아닌 경우 이 이메일을 무시하세요. 감사합니다. ChatGPT 팀 드림 ChatGPT ( https://chatgpt.com ) 도움말 센터 ( https://help.openai.com )",
			expectedCode: "576932",
		},
		{
			name:         "俄文验证码",
			subject:      "Вход в аккаунт",
			body:         "Ваш проверочный код: 491028. Никому не сообщайте этот код.",
			expectedCode: "491028",
		},
		{
			name:         "西班牙语验证码",
			subject:      "Verificación de cuenta",
			body:         "Su código de verificación es: 382910 para continuar.",
			expectedCode: "382910",
		},
		{
			name:         "德语验证码",
			subject:      "Bestätigen Sie Ihre E-Mail-Adresse",
			body:         "Ihr Bestätigungscode lautet: 829103.",
			expectedCode: "829103",
		},
		{
			name:         "法语验证码",
			subject:      "Sécurité du compte",
			body:         "Votre code de confirmation est : 719204 pour activer votre compte.",
			expectedCode: "719204",
		},
		{
			name:         "越南语验证码",
			subject:      "Xác thực tài khoản",
			body:         "Mã xác thực của bạn là: 629105. Hết hạn sau 5 phút.",
			expectedCode: "629105",
		},
		{
			name:         "土耳其语验证码",
			subject:      "Hesap Güvenliği",
			body:         "Doğrulama kodunuz: 839102 olarak belirlenmiştir.",
			expectedCode: "839102",
		},
		{
			name:         "阿拉伯语验证码",
			subject:      "تأكيد الحساب",
			body:         "رمز التحقق الخاص بك هو: 918234 للاستمرار.",
			expectedCode: "918234",
		},
		{
			name:         "日语确认代码",
			subject:      "アカウントの確認",
			body:         "お客様の確認コードは 576932 です。10分以内に入力してください。",
			expectedCode: "576932",
		},
		{
			name:         "标题括号验证码",
			subject:      "[849201] Discord verification code",
			body:         "Hi, here is your one-time verification link or code.",
			expectedCode: "849201",
		},
		{
			name:         "HTML盒式单个空格拆分数字 (5 7 6 9 3 2)",
			subject:      "Uber 登录验证码",
			body:         "<p>您的验证码如下：</p><div class='box'><span>5</span> <span>7</span> <span>6</span> <span>9</span> <span>3</span> <span>2</span></div>",
			expectedCode: "576932",
		},
		{
			name:         "防负向订单号穿透_命中真正验证码",
			subject:      "Order confirmation",
			body:         "Your verification code for order #839201 is 492019. Valid for 10 mins.",
			expectedCode: "492019",
		},
		{
			name:         "Steam Guard 5位字母数字混合码",
			subject:      "Your Steam account: Access from new web or mobile device",
			body:         "Here is the Steam Guard code you need that will allow you to sign in: H8R2K",
			expectedCode: "H8R2K",
		},
		{
			name:         "日文直角引号闭合强标注",
			subject:      "『576932』認証コード",
			body:         "認証コードを入力してください。",
			expectedCode: "576932",
		},
		{
			name:         "花括号强标注",
			subject:      "Security Code",
			body:         "Your account confirmation code is {839201}.",
			expectedCode: "839201",
		},
		{
			name:         "全角数字自动转半角",
			subject:      "安全通知",
			body:         "本次登录验证码为：５７６９３２",
			expectedCode: "576932",
		},
		{
			name:         "HTML内联样式CSS尺寸干扰防护",
			subject:      "Verification code",
			body:         "<style>.security-code { width: 480219px; color: #582910; }</style><p>Your code is: 123456</p>",
			expectedCode: "123456",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			res := ExtractOTP(tt.subject, tt.body)
			if res == nil {
				t.Fatalf("ExtractOTP 返回 nil")
			}
			if tt.expectedCode != "" && res.Code != tt.expectedCode {
				t.Errorf("Code 期望 %s, 实际 %s", tt.expectedCode, res.Code)
			}
			if tt.expectedLink != "" && res.MagicLink != tt.expectedLink {
				t.Errorf("MagicLink 期望 %s, 实际 %s", tt.expectedLink, res.MagicLink)
			}
		})
	}
}

func BenchmarkExtractOTP(b *testing.B) {
	subject := "Your verification code"
	body := "Here is your verification code: 394821 for login. Or click https://example.com/confirm?token=xyz123 to activate."
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = ExtractOTP(subject, body)
	}
}

func TestExtractOTPFalsePositiveImmunity(t *testing.T) {
	// 普通购物/餐厅邮件中包含 shopping(内含 pin) 或 hotpot(内含 otp)，不能误判为验证码
	shoppingMail := "Thank you for shopping: 123456 is your order number."
	if res := ExtractOTP("Receipt", shoppingMail); res != nil && res.Code != "" {
		t.Fatalf("shopping 单词不应被当作 pin 提取验证码，实际得到: %q", res.Code)
	}

	hotpotMail := "Welcome to Hotpot restaurant! Bill: 883921"
	if res := ExtractOTP("Dining", hotpotMail); res != nil && res.Code != "" {
		t.Fatalf("hotpot 单词不应被当作 otp 提取验证码，实际得到: %q", res.Code)
	}
}

File: web/src/utils/sniffer.test.ts
/**
 * [INPUT]: 依赖 utils/sniffer (extractVerifyCode, buildSniffContext, stripHtml, parseSenderInfo)
 * [OUTPUT]: 对外提供前端嗅探器单元测试套件
 * [POS]: web/src/utils 的嗅探逻辑与发件人解析回归防线
 * [PROTOCOL]: 变更时更新此头部，然后检查 CLAUDE.md
 */

import { describe, expect, it } from 'vitest'
import {
  buildSniffContext,
  extractVerifyCode,
  parseSenderInfo,
  stripHtml,
} from './sniffer'

describe('buildSniffContext', () => {
  it('正确联合标题与正文摘要，避免短路吞码', () => {
    expect(buildSniffContext('验证码是 123456', '尊敬的用户您好')).toBe(
      '验证码是 123456 尊敬的用户您好',
    )
    expect(buildSniffContext('普通通知', undefined)).toBe('普通通知')
    expect(buildSniffContext(undefined, '正文内容')).toBe('正文内容')
    expect(buildSniffContext('', '')).toBe('')
  })
})

describe('extractVerifyCode', () => {
  it('提取正向中文上下文验证码', () => {
    expect(extractVerifyCode('您的验证码是：849201，请在 10 分钟内输入。')).toBe('849201')
    expect(extractVerifyCode('本次登录动态码: 1234')).toBe('1234')
    expect(extractVerifyCode('安全码为 98765432')).toBe('98765432')
  })

  it('提取正向英文上下文验证码', () => {
    expect(extractVerifyCode('Your verification code is 394821')).toBe('394821')
    expect(extractVerifyCode('Your code: 582910. Do not share.')).toBe('582910')
    expect(extractVerifyCode('Security PIN is 9283')).toBe('9283')
  })

  it('提取带连字号或空格的验证码', () => {
    expect(extractVerifyCode('Your login code is 123-456')).toBe('123456')
    expect(extractVerifyCode('Your one-time password is: 987 654')).toBe('987654')
  })

  it('提取韩文ChatGPT等外语上下文验证码', () => {
    const koreanMail = '다음 임시 인증 코드를 입력해 계속하세요: 576932 ChatGPT 계정을 생성하고자 하는 것이 본인이 아닌 경우 이 이메일을 무시하세요. 감사합니다. ChatGPT 팀 드림 ChatGPT ( https://chatgpt.com ) 도움말 센터 ( https://help.openai.com )'
    expect(extractVerifyCode(koreanMail)).toBe('576932')
  })

  it('提取全球多语种(俄/西/德/法/越/土/阿/日)验证码及括号直提', () => {
    // 俄语
    expect(extractVerifyCode('Ваш проверочный код: 491028')).toBe('491028')
    // 西班牙语
    expect(extractVerifyCode('Su código de verificación es: 382910')).toBe('382910')
    // 德语
    expect(extractVerifyCode('Ihr Bestätigungscode lautet: 829103')).toBe('829103')
    // 法语
    expect(extractVerifyCode('Votre code de confirmation est : 719204')).toBe('719204')
    // 越南语
    expect(extractVerifyCode('Mã xác thực của bạn là: 629105')).toBe('629105')
    // 土耳其语
    expect(extractVerifyCode('Doğrulama kodunuz: 839102')).toBe('839102')
    // 阿拉伯语
    expect(extractVerifyCode('رمز التحقق الخاص بك هو: 918234')).toBe('918234')
    // 日语
    expect(extractVerifyCode('お客様の確認コードは 576932 です')).toBe('576932')
    // 标题/正文括号强标注
    expect(extractVerifyCode('[849201] Discord code')).toBe('849201')
    expect(extractVerifyCode('【392019】您的验证码')).toBe('392019')
  })

  it('提取反向倒装语序的验证码', () => {
    expect(extractVerifyCode('123456 is your verification code')).toBe('123456')
    expect(extractVerifyCode('839201 为本次确认码')).toBe('839201')
    // 夹带修饰定语
    expect(extractVerifyCode('849201 为您的本次登录验证码，切勿泄露。')).toBe('849201')
    expect(extractVerifyCode('492019 is your AWS verification code. Valid for 10 minutes.')).toBe('492019')
    // 5 位倒装码 (测试非 6 位倒装不漏检)
    expect(extractVerifyCode('49201 is your security code')).toBe('49201')
    // 4 位倒装码
    expect(extractVerifyCode('8392 is your code')).toBe('8392')
  })

  it('标题包含验证码而正文摘要不含时，通过联合上下文成功提取', () => {
    const subject = '[GitHub] Your verification code is 492019'
    const preview = 'Hi Linus, someone tried to sign in to your account. Enter the code above.'
    const context = buildSniffContext(subject, preview)
    expect(extractVerifyCode(context)).toBe('492019')
  })

  it('正文包含 HTML 且仅在正文中含有验证码时，联合上下文剥除 HTML 并成功提取', () => {
    const subject = 'OpenAI 验证邮件'
    const preview = ''
    const body = '<div><p>다음 임시 인증 코드를 입력해 계속하세요: <strong>576932</strong></p></div>'
    const context = buildSniffContext(subject, preview, body)
    expect(extractVerifyCode(context)).toBe('576932')
  })

  it('兜底独立 6 位数字仅在验证类关键词下回退，并排除 2020-2035 年月伪码', () => {
    expect(extractVerifyCode('Notice 202501 has been dispatched')).toBeNull()
    expect(extractVerifyCode('Please enter 582910 to continue')).toBeNull()
    expect(extractVerifyCode('Please enter 582910 to continue, verification required')).toBe('582910')
  })

  it('反向正则不把 barcode/unicode/encode 当验证码', () => {
    expect(extractVerifyCode('Tracking 884421 has a barcode attached')).toBeNull()
    expect(extractVerifyCode('Build 12345 uses unicode')).toBeNull()
    expect(extractVerifyCode('Package 492019 encode complete')).toBeNull()
  })
})

describe('stripHtml & parseSenderInfo', () => {
  it('stripHtml 正确清洗 HTML 标签与样式', () => {
    const raw = '<style>body { color: red; }</style><p>Hello &nbsp; <b>World</b>!</p>'
    expect(stripHtml(raw)).toBe('Hello World !')
  })

  it('parseSenderInfo 提取发件人名字、地址并反解 Apple Relay', () => {
    const info = parseSenderInfo('OpenAI <noreply_at_openai_com_123@icloud.com>')
    expect(info.name).toBe('OpenAI')
    expect(info.email).toBe('noreply_at_openai_com_123@icloud.com')
  })
})
